src/core/tsi/s2a/record_protocol/s2a_crypter.cc

```diff
@@ -598,6 +600,222 @@ grpc_status_code s2a_encrypt(s2a_crypter* crypter, uint8_t* plaintext,
+
+  // Cleanup of auxiliary data.
+  gpr_free(additional_data);
+  gpr_free(nonce);
+
+  if (decrypt_status != GRPC_STATUS_OK) {
+    return INTERNAL_ERROR;
+  }
+  GPR_ASSERT(bytes_written != nullptr);
+  GPR_ASSERT(expected_plaintext_and_record_byte_size == *bytes_written);
+  return OK;
+}
+
+s2a_decrypt_status s2a_decrypt_record(
+    s2a_crypter* crypter, iovec& record_header, const iovec* protected_vec,
+    size_t protected_vec_size, iovec& unprotected_vec, size_t* bytes_written,
+    char** error_details) {
+  GPR_ASSERT(crypter != nullptr);
+  GPR_ASSERT(bytes_written != nullptr);
+  size_t max_plaintext_size;
+  grpc_status_code plaintext_status =
+      s2a_max_plaintext_size(crypter, protected_vec, protected_vec_size,
+                             &max_plaintext_size, error_details);
+  if (plaintext_status != GRPC_STATUS_OK) {
+    return INTERNAL_ERROR;
+  }
+  GPR_ASSERT(static_cast<int>(unprotected_vec.iov_len) >= max_plaintext_size);
+  uint8_t* header = reinterpret_cast<uint8_t*>(record_header.iov_base);
+  const uint16_t wire_version = static_cast<uint16_t>(TLS1_2_VERSION);
+  size_t payload_size = get_total_length(protected_vec, protected_vec_size);
+  size_t tag_size;
+  grpc_status_code tag_status = s2a_tag_size(crypter, &tag_size, error_details);
+  if (tag_status != GRPC_STATUS_OK) {
+    return INTERNAL_ERROR;
+  }
+  if (payload_size >
+      SSL3_RT_MAX_PLAIN_LENGTH + tag_size + /** record type **/ 1) {
+    /** If the plaintext size exceeds the max allowed, the TLS 1.3 RFC demands
+     *  that the record protocol returns a "record overflow" alert and that the
+     *  connection be terminated; for more details, see
+     *  https://tools.ietf.org/html/rfc8446#section-5.2 . **/
+    *error_details = gpr_strdup(kS2ARecordExceedMaxSize);
+    return ALERT_RECORD_OVERFLOW;
+  }
+  size_t expected_payload_size = (header[3] << 8) + header[4];
+  if (record_header.iov_len != SSL3_RT_HEADER_LENGTH ||
+      header[0] != SSL3_RT_APPLICATION_DATA ||
+      header[1] != (wire_version >> 8) || header[2] != (wire_version & 0xff) ||
+      payload_size != expected_payload_size) {
+    *error_details = gpr_strdup(kS2AHeaderIncorrectFormat);
+    return INVALID_RECORD;
+  }
+
+  s2a_decrypt_status decrypt_status = s2a_decrypt_payload(
+      crypter, record_header, protected_vec, protected_vec_size,
+      unprotected_vec, bytes_written, error_details);
+  if (decrypt_status != OK) {
+    return decrypt_status;
+  }
+
+  uint8_t* plaintext = reinterpret_cast<uint8_t*>(unprotected_vec.iov_base);
+  /** Search for the nonzero trailing byte, which encodes the record type and is
+   *  appended to the plaintext. **/
+  size_t i;
+  for (i = *bytes_written - 1; i < *bytes_written; i--) {
+    if (plaintext[i] != 0) {
+      break;
+    }
+  }
+  if (i >= *bytes_written) {
+    *error_details = gpr_strdup(kS2ARecordInvalidFormat);
+    return INVALID_RECORD;
+  }
+  uint8_t record_type = plaintext[i];
+  *bytes_written -= 1;
```

Comment:
I've now changed this line to `*bytes_written = i`, since the for loop determines that the record type byte is at index i.

Reply:
Thanks for fixing this. I think we might still have an off-by-one error. Should we use `i+1`? Otherwise, we're not counting the record_type in the count.

Reasoning: `i` is a zero-based index and `bytes_written` is a counter (IIUC).

Reply:
I think the issue is: what is the caller expecting the value of `bytes_written` to be after the function executes successfully? It should be the number of _plaintext_ bytes written to `unprotected_vec`, not the total number of bytes written. In this case, I think `i` is the number of `plaintext` bytes written, but `i+1` is the total number of bytes written.

Does that make sense? 

If you think this is confusing, I can change the variable name to `plaintext_bytes_written`.

Reply:
That makes sense, thanks for the explanation. +1 to changing the name to make it clearer.